src/iam/certhandler/modules/pkcs11.cpp

```diff
@@ -0,0 +1,986 @@
+    uuid::UUID                      id;
+
+    err = ParseURL(keyURL, label, id);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    const auto privKey = pkcs11::Utils(*session, mAllocator).FindPrivateKey(id, label);
+    if (!privKey.mError.IsNone()) {
+        return privKey.mError;
+    }
+
+    return session->DestroyObject(privKey.mValue.mHandle);
+}
+
+Error PKCS11Module::ValidateCertificates(
+    Array<StaticString<cURLLen>>& invalidCerts, Array<StaticString<cURLLen>>& invalidKeys, Array<CertInfo>& validCerts)
+{
+    Error                             err = Error::Enum::eNone;
+    bool                              isOwned = false;
+    UniquePtr<pkcs11::SessionContext> session;
+
+    Tie(isOwned, err) = IsOwned();
+    if (!err.IsNone() || !isOwned) {
+        return err;
+    }
+
+    Tie(session, err) = CreateSession(true, mUserPIN);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    // search token objects
+    StaticArray<SearchObject, cCertsPerModule> certificates;
+    StaticArray<SearchObject, cCertsPerModule> privKeys;
+    StaticArray<SearchObject, cCertsPerModule> pubKeys;
+
+    SearchObject filter;
+
+    filter.mLabel = mCertType;
+    filter.mType = CKO_CERTIFICATE;
+
+    err = FindObject(*session, filter, certificates);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    filter.mType = CKO_PUBLIC_KEY;
+
+    err = FindObject(*session, filter, privKeys);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    filter.mType = CKO_PRIVATE_KEY;
+
+    err = FindObject(*session, filter, pubKeys);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    // generate valid info
+    err = GetValidInfo(*session, certificates, privKeys, pubKeys, validCerts);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    // create urls for invalid objects
+    err = CreateInvalidURLs(certificates, invalidCerts);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = CreateInvalidURLs(privKeys, invalidKeys);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return CreateInvalidURLs(pubKeys, invalidKeys);
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+String PKCS11Module::GetTokenLabel() const
+{
+    return mConfig.mTokenLabel.IsEmpty() ? cDefaultTokenLabel : mConfig.mTokenLabel;
+}
+
+RetWithError<pkcs11::SlotID> PKCS11Module::GetSlotID()
+{
+    const int paramCount
+        = static_cast<int>(mConfig.mSlotID.HasValue() + mConfig.mSlotIndex.HasValue() + !mConfig.mTokenLabel.IsEmpty());
+
+    if (paramCount > 1) {
+        LOG_ERR()
+            << "Only one parameter for slot identification should be specified (slotId or slotIndex or tokenLabel)";
+
+        return {0U, AOS_ERROR_WRAP(Error::Enum::eInvalidArgument)};
```

Comment:
Why do you need `U` suffix? Do you see any warning? This question relates to all other cases as well.